src/Database/Query.php

```diff
@@ -270,11 +258,18 @@ public function useWriteRole()
      */
     public function useRole(string $role)
     {
-        if ($role === Connection::ROLE_READ) {
-            return $this->useReadRole();
+        assert(in_array($role, [Connection::ROLE_READ, Connection::ROLE_WRITE], true));
+        if ($this->_connection->role() === $role) {
+            return $this;
+        }
+
+        $name = $this->_connection->configName();
+        $roleName = ConnectionManager::getName($role, $name);
+        if ($roleName === $name) {
+            return $this;
         }
 
-        return $this->useWriteRole();
+        return $this->setConnection(ConnectionManager::get($roleName));
```

Comment:
We check the new name against the current name before switching to support manual write connections.

If someone manually created a read connection `users:read` and tries to switch to the write role, it will fail as expected still.